sw/device/tests/spi_device_ujson_ottf_test.c

```diff
@@ -0,0 +1,48 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/base/status.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/json/command.h"
+#include "sw/device/lib/testing/spi_device_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_console.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/test_framework/ujson_ottf.h"
+#include "sw/device/lib/ujson/ujson.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG(.console.type = kOttfConsoleSpiDevice,
+                        .console.base_addr = TOP_EARLGREY_SPI_DEVICE_BASE_ADDR,
+                        .console.test_may_clobber = false, );
+
+status_t dummy_sw_strap_read(ujson_t *uj) {
```

Comment:
This test has nothing to do with SW straps correct? If you need some UJSON Structs to send back and for testing, why don't you use this generic `MemRead` test command https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/json/command.h;l=25 , and send MemRead data payloads back and forth for one test: https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/lib/testing/json/mem.h;drc=30d7e787c753caaa03fe68a4a70da1bbcbc1d96f;l=24 (this should test data payloads of greater than 256 bytes).

Then you could also test sending a perso blob ujson struct to and from the host since it is > 4k in size which should excercise multi SPI buffer window transactions.

Also if you could add a block comment to the top describing what you are testing that would be helpful too.

Reply:
Yes, the test has nothing to do with the functionality of SW straps. I did consider using the generic `Mem*` test commands but decided to use the SW straps command in the end. The tx transaction on SPI device console requires the device to send a sync message to notify the host it's ready for transmission but this limitation doesn't exist on UART console. The `mem` module encapsulates several tx transactions into a single `execute` function. To make the mem module support the workflow SPI device console, the existing `example_mem_ujcmd` test would be affected as well. I've submitted a new commit for refactoring the mem module. Let me know if that works for you.   I also added a test to send a perso blob ujson struct back and forth, thanks for the suggestion!

Reply:
Thanks for the comment. The purpose is to ensure that the rx data has been read  by the host before host starts sending data to the device, so the device won't be blocked by the rx transmission if the read buffer is full and can be ready to receive the tx data. However, after I think about the use case here again, the sync operation in the `execute` seems to be unnecessary if a sync operation has already been performed in advance. Will double confirm on this and see if I can get rid of the `require_sync`

Reply:
Maybe this is a limitation of opentitanlib but I would have thought that the console would have a background thread so it can periodically read the SPI and therefore unblock the device if necessary?

Reply:
> The purpose is to ensure that the rx data has been read by the host before host starts sending data to the device, so the device won't be blocked by the rx transmission if the read buffer is full and can be ready to receive the tx data.

Couldn't the host just poll until the rx buffer is empty before sending data out? Would the device be sending data unsolicited to the host?

Reply:
Removed the usage of `require_sync` in the new commit. Host poll until the rx buffer is empty before sending data out should work as well. Ideally, the device should not send unsolicited data to the host. I observed the issue when I tried to log plenty of debug messages on the device side but didn't handle those messages properly on the host side. After checking the behaviors on the existing ottf test code, I think the case I observed is an extreme case that should not occur in the real use scenario. The original purpose is not really necessary now.